### PR TITLE
test and fix CreateIdTransaction bug

### PR DESCRIPTION
**Pull request by ttblack**

### CODE REVIEW

File: SDK/Implement/SubWallet.cpp [hunk 1]
```diff
@@ -224,6 +225,25 @@ namespace Elastos {
 			return j;
 		}
 
+		void SubWallet::removeDuplicatedPrograms(const TransactionPtr &transaction) {
+			std::vector<Program *> &programs = (std::vector<Program *> &)transaction->getPrograms();
```
Comment (raindust) [inline SDK/Implement/SubWallet.cpp]: Use "const std::vector<Program *> &" here directly

File: SDK/Implement/SubWallet.cpp [hunk 2]
```diff
@@ -224,6 +225,25 @@ namespace Elastos {
...
+			std::map<std::string, Program *> programMap;
+			size_t len = programs.size();
+			for (size_t i = 0; i < len; ++i) {
+				std::string key = Utils::encodeHex(programs[i]->getCode());
+				if (programMap.find(key) == programMap.end()) {
+					programMap[key] = programs[i];
+				}
+			}
+			programs.clear();
```
Comment (raindust) [inline SDK/Implement/SubWallet.cpp]: Ok, let it be, but add a "clearPrograms" will be more elegant.

File: SDK/Implement/SubWallet.h [hunk 3]
```diff
@@ -161,6 +161,7 @@ namespace Elastos {
 													  const std::string &status,
 													  const nlohmann::json &desc,
 													  uint32_t confirms);
+			void removeDuplicatedPrograms(const TransactionPtr &transaction);
```
Comment (raindust) [inline SDK/Implement/SubWallet.h]: This method can be protected

Comment (ttblack) [reply SDK/Implement/SubWallet.h]: this method is already protected

File: SDK/Transaction/Transaction.cpp [hunk 4]
```diff
@@ -337,6 +344,9 @@ namespace Elastos {
 			return _transaction->programs;
 		}
 
+		void Transaction::clearPrograms() {
+			_transaction->programs.clear();
```
Comment (heropan) [inline SDK/Transaction/Transaction.cpp]: Here can not just clear(), it will cause memory leak.

File: SDK/Implement/SubWallet.cpp [hunk 5]
```diff
@@ -224,6 +225,26 @@ namespace Elastos {
...
 		}
 
+		void SubWallet::removeDuplicatedPrograms(const TransactionPtr &transaction) {
+			const std::vector<Program *> &programs = transaction->getPrograms();
+			std::map<std::string, Program *> programMap;
+			size_t len = programs.size();
+			for (size_t i = 0; i < len; ++i) {
+				std::string key = Utils::encodeHex(programs[i]->getCode());
+				if (programMap.find(key) == programMap.end()) {
+					programMap[key] = programs[i];
```
Comment (heropan) [inline SDK/Implement/SubWallet.cpp]: memory leak here too...

File: SDK/Implement/SubWallet.h (hunk 3 above)
Comment (heropan) [reply SDK/Implement/SubWallet.h]: Move this function to class Transaction is better ?

**Comment by ttblack** [comment]
@raindust @heropan moidfy removeDuplicatePrograms to Trasaction.cpp and modify logic ,please review again.

**Pull request merged**